src/rand.rs

```diff
@@ -123,23 +120,87 @@ use self::sysrand_or_urandom::fill as fill_impl;
 #[cfg(any(target_os = "macos", target_os = "ios"))]
 use self::darwin::fill as fill_impl;
 
+#[cfg(target_os = "linux")]
+mod sysrand_chunk {
+    use error;
+    use c;
+
+    extern {
+        fn syscall(number: c::long, ...) -> c::long;
+        fn GFp_errno() -> c::int;
+    }
+
+    #[cfg(target_arch = "aarch64")]
+    const GETRANDOM: c::long = 278;
+    #[cfg(target_arch = "arm")]
+    const GETRANDOM: c::long = 384;
+    #[cfg(target_arch = "x86")]
+    const GETRANDOM: c::long = 355;
+    #[cfg(target_arch = "x86_64")]
+    const GETRANDOM: c::long = 318;
+
+    const EINTR: c::int = 4;
+
+    #[inline]
+    pub fn chunk(dest: &mut [u8]) -> Result<usize, error::Unspecified> {
+        let chunk_len: c::size_t = dest.len();
+        let flags: c::uint = 0;
+        let r = unsafe {
+            syscall(GETRANDOM, dest.as_mut_ptr(), chunk_len, flags)
+        };
+        if r < 0 {
+            if unsafe { GFp_errno() } == EINTR {
+                // If an interrupt occurs while getrandom() is blocking
+                // to wait for the entropy pool, then EINTR is returned.
+                // Returning 0 will cause the caller to try again.
+                return Ok(0);
+            }
+            return Err(error::Unspecified);
+        }
+        Ok(r as usize)
+    }
+}
+
+#[cfg(windows)]
+mod sysrand_chunk {
+    use std::cmp::min;
+    use error;
+    use c::win32::{ULONG, BOOLEAN};
+    use core::mem::size_of;
+
+    #[link(name = "Advapi32")]
+    extern "system" {
+        // also known as RtlGenRandom
+        fn SystemFunction036(random_buffer: *mut u8,
+                             random_buffer_length: ULONG)
+                             -> BOOLEAN;
+    }
+
+    #[inline]
+    pub fn chunk(dest: &mut [u8]) -> Result<usize, error::Unspecified> {
+        assert!(size_of::<usize>() >= size_of::<ULONG>());
+        let max_chunk_len = (1 << (size_of::<ULONG>()*8 - 1)) - 1;
+        assert_eq!(max_chunk_len, 2147483647);
```

Comment:
I did that because the old code enforced that bound: https://github.com/briansmith/ring/blob/48f6b36582a0743d4c9678949efee000c6209d34/crypto/rand/sysrand.c#L69 

I could not find any documentation to support needing to have that limit, though.